tracing/traced_file.cc

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (C) 2019 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <seastar/core/future.hh>
+#include <seastar/core/print.hh>
+
+#include "tracing/traced_file.hh"
+
+using namespace seastar;
+
+class traced_file_impl : public file_impl {
+private:
+    // Invariant: _f_impl == *get_file_impl(_f).
+    // The _f is only stored to keep ownership.
+    file _f;
+    file_impl& _f_impl;
+
```

Comment:
Why store _f_impl?

Reply:
So that I don't call `get_file_impl(_f)` every time I want to perform an operation.

I would ask the other question instead: why store `_f`? `_f_impl` is really what I need. But I must store `_f` because it manages `_f_impl`'s lifetime.

Reply:
It's an inline that translates to the same thing as dereferencing a reference.

I knew why you stored _f.